torch/_dynamo/polyfills/tensor.py

```diff
@@ -15,7 +15,10 @@ def make_subclass(
     # suffice for most of Dynamo tracing purposes.
     # https://github.com/pytorch/pytorch/blob/ccfde4dadfa3c342076a1ee387017f84dd4ad2f7/torch/csrc/autograd/python_variable.cpp#L597-L650
     assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
-    data = data.detach()
+
+    # Avoid unnecessary `detach` node in the graph.
+    if data.requires_grad:
+        data = data.detach()
```

Comment:
https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/csrc/autograd/python_variable.cpp#L617-L618

make_subclass unconditionally applies detach. Why should we avoid doing this when requires_grad is False? I'm not convinced the semantics are the same.

Reply:
> Why should we avoid doing this when requires_grad is False?

Mainly because it gets added aggressively into the Dynamo graph and increases AOTDispatcher compilation time _quite_ a bit: #150706, which might be caused by #151142 but I'm not sure.

I also thought about using `alias` by apparently the semantics aren't the same either:
```
>>> import torch
>>> x = torch.ones(2)
>>> x_detach = x.detach()
>>> x_alias = torch.ops.aten.alias(x)
>>> x += 1
>>> x_detach
tensor([2., 2.])
>>> x_alias
tensor([2., 2.])
>>> x_detach._is_view()
False
>>> x_alias._is_view()
True
```

Reply:
Do we have to perfectly emulate the [detach semantics](https://github.com/pytorch/pytorch/blob/d5a8bd068821ca0cd036400c13a561452d391b15/aten/src/ATen/native/TensorShape.cpp#L4504-L4512) here though? In fact before I added this `detach` emulation in #149483, we never even emulated it at all.

The detach call under `requires_grad=False` looks mostly like a super shallow copy (not even view), and since the detached data will be fed into a `as_subclass` or `torch.Tensor` call below (which is another shallow copy), I wonder if there's any way for the end user to observe the differences of a perfect `detach `emulation v.s. not in real code

Reply:
I think I am fine as long as the output tensor of the _make_subclass call is not a view of the input tensor (when requires_grad=False). I don't know how to trigger this code, if you have a repro, the view check looks like:
```
x = ... (input)
y = ...  # the make_subclass call
assert y._base is None
```